client/utils/rest.go

```diff
@@ -68,3 +87,134 @@ func WriteGenerateStdTxResponse(w http.ResponseWriter, txBldr authtxb.TxBuilder,
+
+//----------------------------------------
+// Building / Sending utilities
+
+type BaseReq struct {
+	Name          string `json:"name"`
+	Password      string `json:"password"`
+	ChainID       string `json:"chain_id"`
+	AccountNumber int64  `json:"account_number"`
+	Sequence      int64  `json:"sequence"`
+	Gas           int64  `json:"gas"`
+	GasAdjustment string `json:"gas_adjustment"`
+}
+
+/*
+ReadRESTReq is a simple convenience wrapper that reads the body and
+unmarshals to the req interface.
+
+  Usage:
+    type SomeReq struct {
+      BaseReq            `json:"base_req"`
+      CustomField string `json:"custom_field"`
+    }
+    req := new(SomeReq)
+    err := ReadRESTReq(w, r, cdc, req)
+*/
+func ReadRESTReq(w http.ResponseWriter, r *http.Request, cdc *wire.Codec, req interface{}) error {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+		return err
+	}
+	err = cdc.UnmarshalJSON(body, req)
+	if err != nil {
+		WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (req BaseReq) BaseReqValidate(w http.ResponseWriter) bool {
+	if len(req.Name) == 0 {
+		WriteErrorResponse(w, http.StatusUnauthorized, "Name required but not specified")
+		return false
+	}
+
+	if len(req.Password) == 0 {
+		WriteErrorResponse(w, http.StatusUnauthorized, "Password required but not specified")
+		return false
+	}
+
+	if len(req.ChainID) == 0 {
+		WriteErrorResponse(w, http.StatusUnauthorized, "ChainID required but not specified")
+		return false
+	}
+
+	if req.AccountNumber < 0 {
+		WriteErrorResponse(w, http.StatusUnauthorized, "Account Number required but not specified")
+		return false
+	}
+
+	if req.Sequence < 0 {
+		WriteErrorResponse(w, http.StatusUnauthorized, "Sequence required but not specified")
+		return false
+	}
+	return true
+}
+
+// CompleteAndBroadcastTxREST implements a utility function that
+// facilitates sending a series of messages in a signed
+// transaction given a TxBuilder and a QueryContext. It ensures
+// that the account exists, has a proper number and sequence
+// set. In addition, it builds and signs a transaction with the
+// supplied messages.  Finally, it broadcasts the signed
+// transaction to a node.
+// NOTE: Also see CompleteAndBroadcastTxCli.
+// NOTE: Also see x/stake/client/rest/tx.go delegationsRequestHandlerFn.
+func CompleteAndBroadcastTxREST(w http.ResponseWriter, r *http.Request, cliCtx context.CLIContext, baseReq BaseReq, msgs []sdk.Msg, cdc *wire.Codec) {
+	var err error
+	txBldr := authtxb.TxBuilder{
+		Codec:         cdc,
+		AccountNumber: baseReq.AccountNumber,
+		Sequence:      baseReq.Sequence,
+		ChainID:       baseReq.ChainID,
+		Gas:           baseReq.Gas,
+	}
+
+	adjustment, ok := ParseFloat64OrReturnBadRequestDefault(w, baseReq.GasAdjustment, client.DefaultGasAdjustment)
+	if !ok {
+		return
+	}
+	cliCtx = cliCtx.WithGasAdjustment(adjustment)
```

Comment:
why are we using `float64` here?

Reply:
It's the multiplicative factor applied to the gas estimate, e.g. estimate = 2000, adjustment = 1.1 => adjusted = 2200

Reply:
Can we use `sdk.Dec` for consistency?

Reply:
let's use `sdk.Dec`

Reply:
Will address in #2400 as I think the PR (#2391) is already large enough.